# **Exploring eBay Car Sales Data** 
This notebook will be destroyed/deleted at the end.

## **Table of contents**

1. [Introduction]()
<ul>
    <li>Project Description</li>
    <li>Data Description</li>
</ul>

2. [Importing and Loading Data]()
<ul>
    <li>
        Importing Libraries
    </li>
    <li>
        Reading in Data
    </li>
    <li>
        Exploring the Data
    </li>
</ul>

3. [Data Cleaning]()
<ul>
    <li>Cleaning Column Names</li>    
</ul>

4. [Data Analysis]()
<ul>
    <li>Initial Exploration and Cleaning</li>
    <ul>
        <li>Candidate columns to be dropped</li>
        <li>Columns that need more investigation</li>
        <li>Columns that need dtype conversion</li>
    </ul>
</ul>


## **Introduction**

### Project Description

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

### Data Description

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeen` - When the crawler saw this ad last online.


## **Importing and Loading Data**

### Importing Libraries

In [1]:
import pandas as pd

### Reading in Data

In [2]:
autos = pd.read_csv('../Datasets/autos.csv', encoding = 'Latin-1')

### Exploring the Data

In [3]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [5]:
autos.describe(include = 'all')

dateCrawled         name  seller offerType         price  \
count                371528       371528  371528    371528  3.715280e+05   
unique               280500       233531       2         2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat   Angebot           NaN   
freq                      7          657  371525    371516           NaN   
mean                    NaN          NaN     NaN       NaN  1.729514e+04   
std                     NaN          NaN     NaN       NaN  3.587954e+06   
min                     NaN          NaN     NaN       NaN  0.000000e+00   
25%                     NaN          NaN     NaN       NaN  1.150000e+03   
50%                     NaN          NaN     NaN       NaN  2.950000e+03   
75%                     NaN          NaN     NaN       NaN  7.200000e+03   
max                     NaN          NaN     NaN       NaN  2.147484e+09   

        abtest vehicleType  yearOfRegistration  gearbox        powerPS  \
count   371528      333659       371528.000000   351319  371528.000000   
unique       2           8                 NaN        2            NaN   
top       test   limousine                 NaN  manuell            NaN   
freq    192585       95894                 NaN   274214            NaN   
mean       NaN         NaN         2004.577997      NaN     115.549477   
std        NaN         NaN           92.866598      NaN     192.139578   
min        NaN         NaN         1000.000000      NaN       0.000000   
25%        NaN         NaN         1999.000000      NaN      70.000000   
50%        NaN         NaN         2003.000000      NaN     105.000000   
75%        NaN         NaN         2008.000000      NaN     150.000000   
max        NaN         NaN         9999.000000      NaN   20000.000000   

         model      kilometer  monthOfRegistration fuelType       brand  \
count   351044  371528.000000        371528.000000   338142      371528   
unique     251            NaN                  NaN        7          40   
top       golf            NaN                  NaN   benzin  volkswagen   
freq     30070            NaN                  NaN   223857       79640   
mean       NaN  125618.688228             5.734445      NaN         NaN   
std        NaN   40112.337051             3.712412      NaN         NaN   
min        NaN    5000.000000             0.000000      NaN         NaN   
25%        NaN  125000.000000             3.000000      NaN         NaN   
50%        NaN  150000.000000             6.000000      NaN         NaN   
75%        NaN  150000.000000             9.000000      NaN         NaN   
max        NaN  150000.000000            12.000000      NaN         NaN   

       notRepairedDamage          dateCreated  nrOfPictures    postalCode  \
count             299468               371528      371528.0  371528.00000   
unique                 2                  114           NaN           NaN   
top                 nein  2016-04-03 00:00:00           NaN           NaN   
freq              263182                14450           NaN           NaN   
mean                 NaN                  NaN           0.0   50820.66764   
std                  NaN                  NaN           0.0   25799.08247   
min                  NaN                  NaN           0.0    1067.00000   
25%                  NaN                  NaN           0.0   30459.00000   
50%                  NaN                  NaN           0.0   49610.00000   
75%                  NaN                  NaN           0.0   71546.00000   
max                  NaN                  NaN           0.0   99998.00000   

                   lastSeen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

## **Data Cleaning**

###  Cleaning Column Names

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
new_list = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns = new_list
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

The following edits to columns names have been done in the previous code blocks:

- yearOfRegistration to registration_year
- monthOfRegistration to registration_month
- notRepairedDamage to unrepaired_damage
- dateCreated to ad_created
- The rest of the column names from camelcase to snakecase.

The pandas convention for column names is snakecase rather than camel casing which came with the raw dataset. We have made the changes according to this pandas convention for easy understanding and convenience. Other column names have also been changed for better readibility and thus to improve productivity.

### Converting categorical data that uses German words to English.

In [8]:
# Identifying categorical data that uses German words and translate them to their English meaning.
autos['seller'] = autos['seller'].map({'privat': 'private', 'gewerblich': 'commercial'})
autos['offer_type'] = autos['offer_type'].map({'Angebot': 'offer','Gesuch':'request'})
autos['gearbox'] = autos['gearbox'].map({'manuell': 'manual', 'automatik': 'automatic'})
autos['unrepaired_damage'] = autos['unrepaired_damage'].map({'nein': 'no', 'ja':'yes'})
autos['fuel_type'] = autos['fuel_type'].map({'benzin': 'benzine', 'andere' : 'other', 'elektro': 'electro'})

### Converting the dates to be numeric


In [9]:
autos['date_crawled'].str[:10].value_counts(dropna = False).sort_index()

date_crawled
2016-03-05     9493
2016-03-06     5373
2016-03-07    13260
2016-03-08    12429
2016-03-09    12710
2016-03-10    12098
2016-03-11    12159
2016-03-12    13447
2016-03-13     5846
2016-03-14    13477
2016-03-15    12428
2016-03-16    11201
2016-03-17    11763
2016-03-18     4877
2016-03-19    13112
2016-03-20    13506
2016-03-21    13278
2016-03-22    12063
2016-03-23    11878
2016-03-24    11113
2016-03-25    12236
2016-03-26    11876
2016-03-27    11248
2016-03-28    13045
2016-03-29    12695
2016-03-30    12457
2016-03-31    11846
2016-04-01    12675
2016-04-02    13033
2016-04-03    14391
2016-04-04    13974
2016-04-05     4764
2016-04-06     1175
2016-04-07      602
Name: count, dtype: int64

In [10]:
# Below code will convert the date in the 

## **Data Analysis**

### Initial Exploration and Cleaning

In [11]:
autos.dtypes

date_crawled          object
name                  object
seller                object
offer_type            object
price                  int64
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
kilometer              int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object

In [12]:
autos.describe(include = 'all')        

date_crawled         name   seller offer_type         price  \
count                371528       371528   371528     371528  3.715280e+05   
unique               280500       233531        2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  private      offer           NaN   
freq                      7          657   371525     371516           NaN   
mean                    NaN          NaN      NaN        NaN  1.729514e+04   
std                     NaN          NaN      NaN        NaN  3.587954e+06   
min                     NaN          NaN      NaN        NaN  0.000000e+00   
25%                     NaN          NaN      NaN        NaN  1.150000e+03   
50%                     NaN          NaN      NaN        NaN  2.950000e+03   
75%                     NaN          NaN      NaN        NaN  7.200000e+03   
max                     NaN          NaN      NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year gearbox       power_ps   model  \
count   371528       333659      371528.000000  351319  371528.000000  351044   
unique       2            8                NaN       2            NaN     251   
top       test    limousine                NaN  manual            NaN    golf   
freq    192585        95894                NaN  274214            NaN   30070   
mean       NaN          NaN        2004.577997     NaN     115.549477     NaN   
std        NaN          NaN          92.866598     NaN     192.139578     NaN   
min        NaN          NaN        1000.000000     NaN       0.000000     NaN   
25%        NaN          NaN        1999.000000     NaN      70.000000     NaN   
50%        NaN          NaN        2003.000000     NaN     105.000000     NaN   
75%        NaN          NaN        2008.000000     NaN     150.000000     NaN   
max        NaN          NaN        9999.000000     NaN   20000.000000     NaN   

            kilometer  registration_month fuel_type       brand  \
count   371528.000000       371528.000000    224169      371528   
unique            NaN                 NaN         3          40   
top               NaN                 NaN   benzine  volkswagen   
freq              NaN                 NaN    223857       79640   
mean    125618.688228            5.734445       NaN         NaN   
std      40112.337051            3.712412       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                   no  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

In [13]:
autos.describe(include = 'O')

date_crawled         name   seller offer_type  abtest  \
count                371528       371528   371528     371528  371528   
unique               280500       233531        2          2       2   
top     2016-03-24 14:49:47  Ford_Fiesta  private      offer    test   
freq                      7          657   371525     371516  192585   

       vehicle_type gearbox   model fuel_type       brand unrepaired_damage  \
count        333659  351319  351044    224169      371528            299468   
unique            8       2     251         3          40                 2   
top       limousine  manual    golf   benzine  volkswagen                no   
freq          95894  274214   30070    223857       79640            263182   

                 ad_created            last_seen  
count                371528               371528  
unique                  114               182806  
top     2016-04-03 00:00:00  2016-04-07 06:45:59  
freq                  14450                   17

In [14]:
autos.describe()

price  registration_year       power_ps      kilometer  \
count  3.715280e+05      371528.000000  371528.000000  371528.000000   
mean   1.729514e+04        2004.577997     115.549477  125618.688228   
std    3.587954e+06          92.866598     192.139578   40112.337051   
min    0.000000e+00        1000.000000       0.000000    5000.000000   
25%    1.150000e+03        1999.000000      70.000000  125000.000000   
50%    2.950000e+03        2003.000000     105.000000  150000.000000   
75%    7.200000e+03        2008.000000     150.000000  150000.000000   
max    2.147484e+09        9999.000000   20000.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count       371528.000000        371528.0  371528.00000  
mean             5.734445             0.0   50820.66764  
std              3.712412             0.0   25799.08247  
min              0.000000             0.0    1067.00000  
25%              3.000000             0.0   30459.00000  
50%              6.000000             0.0   49610.00000  
75%              9.000000             0.0   71546.00000  
max             12.000000             0.0   99998.00000

#### **Columns that have mostly one value that are candidates to be dropped.**

- `seller:` This column has `private` as the one value occurring (371525	out of 371528 times) - 99.99%
-  `offer_type:` This column has `offer` occurring (371516 out of 371528 times) - 99.99%

The above columns have mostly one value. Also, we see some of them lacking entries and have other values in higher majority.

However, we see the `seller` and `offer_type` columns are suitable candidates to be dropped as they may not contribute to valuable inferrences. They just have one value occurring more than 99% of the time.

In [15]:
autos['seller'].value_counts(dropna = False) # This will gives us the other minority value in this column.

seller
private       371525
commercial         3
Name: count, dtype: int64

We see there are just two values present in this column.

In [16]:
autos['offer_type'].value_counts()

offer_type
offer      371516
request        12
Name: count, dtype: int64

This column is similar to the `seller` column with a high majority value and a minor one.

These columns can be dropped.

In [17]:
autos['fuel_type'].value_counts(dropna = False)

fuel_type
benzine    223857
NaN        147359
other         208
electro       104
Name: count, dtype: int64

In [18]:
autos['unrepaired_damage'].value_counts(dropna = False)

unrepaired_damage
no     263182
NaN     72060
yes     36286
Name: count, dtype: int64

In [19]:
autos['abtest'].value_counts(dropna = False)

abtest
test       192585
control    178943
Name: count, dtype: int64

#### **Columns that need more investigation.**

- The column `price` might require further investigation, we see the min price being shown as 0.00 and a max being more than a 9-digit number. This column indicates prices for used cars.
- The column `registration_year` requires further investigation. The min year of registration shown, as 1000 and max being shown as 9999. These numbers do not integrate well with the essence of our dataset.
- `power_ps` column has a min value of 0.00. This column requires further investigation.
- `vehicle_type` column has 37869 (371528 - 333659) NaN values, requires further investigation.
- `model` column has 20484 (371528 - 351044) NaN values, requires further investigation.
- `kilometer` column requires further investigation, the max values seem to represent extreme unusual values.
- `fuel_type` column has 33386 NaN values, requires further investigation.
- `registration_month` has min month as 0.00, this requires further investigation.
- `nr_of_pictures` column has mostly zeroes as its elements. This column is a candidate to be dropped.
-  `abtest:` This column has `test` occurring (192585 /371528) times - 51.83%
-  `gearbox:` This column has `manual` occurring (274214 / 351319) times - 78.05% with lacking 20209 entries when compared with the majority of the columns, requires further investigation.
-  `unrepaired_damage:` This column has `no` occurring (263182 / 299468) times - 87.88% with lacking 72060 entries when compared with majority of the columns, requires further investigation.

#### **Columns of numeric data stored as text**

There are no numeric columns stored as text dtype.

#### `postal_code`

In [20]:
autos['postal_code'].value_counts(dropna = False).sort_index()

postal_code
1067     96
1068      1
1069     59
1097     29
1099     67
         ..
99988     9
99991     2
99994     7
99996     3
99998    16
Name: count, Length: 8150, dtype: int64

In [21]:
autos['postal_code'].value_counts(dropna = False).head(10)

postal_code
10115    828
65428    637
66333    349
38518    326
44145    323
32257    323
52525    314
78224    309
26789    301
48599    294
Name: count, dtype: int64

Germany has postal code ranging for (1067 - 99998). This column reflects revelant genuine data.

We see that the postal code, 10115 as the location for most cars listings. With almost 800 listings.

This makes sense as this is the postal code for the capital Berlin.

#### `price`

In [22]:
autos['price'].value_counts(dropna = False).sort_index(ascending = False) #sorts the index in descending order.

price
2147483647        1
99999999         15
99000000          1
74185296          1
32545461          1
              ...  
4                 1
3                 8
2                12
1              1189
0             10778
Name: count, Length: 5597, dtype: int64

There seems to be prices higher than normal for used resold cars. 

Numbers ranging in more than 6 figures, on the other end the some price values are very low, with 0 being the most common.

In [23]:
autos['price'].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

The `price` column has a relatively higher standard deviation, which indicates the data not being closely distributed with the mean.
The reasons for this can be outliers, or the extreme data points in the `price` column, like the high and low price values.

A closely distributed data represents a normal distribution.

In [24]:
autos['price'].value_counts().head()

price
0       10778
500      5670
1500     5394
1000     4649
1200     4594
Name: count, dtype: int64

In [25]:
autos['price'].value_counts().tail()

price
349000    1
8889      1
3440      1
1997      1
10985     1
Name: count, dtype: int64

In [26]:
autos = autos[autos['price'].between(1, 350000)] # filters the dataframe to only include rows where the price columns's value range from 1 to 350000.

In [27]:
autos['price'].describe()

count    360635.000000
mean       5898.671956
std        8866.359669
min           1.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [28]:
autos['price'].value_counts().head()

price
500     5670
1500    5394
1000    4649
1200    4594
2500    4438
Name: count, dtype: int64

In [29]:
autos['price'].value_counts().tail()

price
17501    1
12567    1
31555    1
7675     1
10985    1
Name: count, dtype: int64

In [30]:
autos['price'].value_counts().sort_index().tail(10) # sorts the index in ascending order and returns the bottom 10 values.

price
284000    1
285000    1
294900    1
295000    1
299000    3
300000    1
323223    1
345000    1
349000    1
350000    4
Name: count, dtype: int64

In [31]:
autos['price'].value_counts().sort_index().head(10)

price
1     1189
2       12
3        8
4        1
5       26
7        3
8        9
9        8
10      84
11       5
Name: count, dtype: int64

In [32]:
autos = autos[autos['price'].between(100, 100000)]

In [33]:
autos['price'].describe()

count    357805.000000
mean       5816.364481
std        7562.614881
min         100.000000
25%        1290.000000
50%        3100.000000
75%        7499.000000
max      100000.000000
Name: price, dtype: float64

We see now, the mean shows a reasonable value, and the standard deviation is at a lower value than before. This supports the distribution, to be more closely represented amongst the mean.
The minimum and maximum values are now changed. After removing the outliers in the column, the data is closely represents a normal distribution. 

An higher standard deviation represents uncertainty in data and a distribution not closely related to normal.

We see that the average price of a car is 5816.36.
- The 25th percentile shows 25% of the cars listed has price less than or equal to 1290.
- The 50th percentile, indicates 50% have price less than or equal to 3100.
- The 75th percentile is to show that 75% have price less than or equal to 7499.
We see the max price is 100000.

In [34]:
autos['price'].isnull().sum() #counts the number of null values in a column.

0

In [35]:
autos['price'].kurtosis()

20.986781043874668

The price distribution here is shown to be leptokurtic with thicker tails and value closer to the mean. This distribution is seen after removal of the outliers.

#### `registration_year`

In [36]:
autos['registration_year'].value_counts().sort_index() #Using series.value_counts() and then series.sort_index(), brings the values to the index side and sorts them. 

registration_year
1000    22
1001     1
1039     1
1111     1
1234     4
        ..
8500     1
8888     1
9000     4
9450     1
9999    13
Name: count, Length: 143, dtype: int64

We see, above that there seems to be unrealistic min years less than the 90's or 80's, even before cars were invented and the resold site was up and running. There also seems to be incorrect data of years, being in the future.

We will go ahead and create a window to remove outliers and uncertain data.

In [37]:
autos['registration_year'].describe()

count    357805.000000
mean       2004.294445
std          72.258264
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

We see that the standard deviation is further from the mean, indicating again that the data is not evenly spread out or closely distributed amongst the mean.
This thereby indicates uncertainty in the dataset.

In [38]:
autos['registration_year'].value_counts().head(20)

registration_year
2000    22706
1999    21881
2005    21562
2006    19962
2001    19658
2003    19558
2004    19430
2002    18769
2007    17459
1998    17067
2008    15995
2009    15471
1997    13879
2010    12226
2011    11964
1996    10297
2017     9935
2012     9334
2016     9052
1995     8746
Name: count, dtype: int64

In [39]:
autos['registration_year'].value_counts().tail(20)

registration_year
7100    1
1400    1
8888    1
2200    1
5900    1
2900    1
8500    1
5300    1
1255    1
2066    1
3700    1
3800    1
4800    1
1911    1
2290    1
4100    1
1001    1
6200    1
9450    1
8200    1
Name: count, dtype: int64

In [40]:
autos['registration_year'].value_counts().sort_index().head(20)

registration_year
1000    22
1001     1
1039     1
1111     1
1234     4
1255     1
1300     2
1400     1
1500     3
1600     2
1602     1
1800     4
1910    36
1911     1
1923     3
1925     1
1927     2
1928     2
1929     7
1930     4
Name: count, dtype: int64

In [41]:
autos = autos[autos['registration_year'].between(1900, 2020)]
autos['registration_year'].describe()

count    357683.000000
mean       2003.482184
std           7.549755
min        1910.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        2019.000000
Name: registration_year, dtype: float64

We know see the min and max values having significant realistic values relating to the present's timeline. The SD has also significantly decreased after the removal of outliers.

In [42]:
autos['registration_year'].value_counts(dropna = False).sort_index()

registration_year
1910      36
1911       1
1923       3
1925       1
1927       2
        ... 
2015    2817
2016    9052
2017    9935
2018    3798
2019      16
Name: count, Length: 97, dtype: int64

In [43]:
autos['registration_year'].isnull().sum()

0

In [44]:
autos['registration_year'].kurtosis()

7.54898523472132

In [45]:
autos['registration_year'].describe()

count    357683.000000
mean       2003.482184
std           7.549755
min        1910.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        2019.000000
Name: registration_year, dtype: float64

The average year of registration is 2004, this indicates that cars being sold are within the last two decades.

Where the average vehicle sold is 19 - 20 years old.

We also see that,
- The 25th percentile is 2000, meaning 25% of cars were registered before 2000.
- The 50th percentile is 2005, suggesting that 50% were registered before 2005.
- The 75th percentile is 2009, indicating that 75% of the cars were registered before 2009.

While the maximum year of registration after removal of outliers is seen as 2019.

#### `power_ps`

In [46]:
autos['power_ps'].describe()

count    357683.000000
mean        116.716333
std         188.905782
min           0.000000
25%          73.000000
50%         107.000000
75%         150.000000
max       20000.000000
Name: power_ps, dtype: float64

The `power_ps` columns describe the power of the car in PS.

PS or PferdStarke is the metric measure of horsepower. It is the equivalent of 98.6% of one HP.

For midsize cars, 250 is a good start. Anything under 180 might be slightly underpowered unless the vehicle is incredibly small like a Mini Cooper or Mitsubishi Mirage. 

 For example, 150 PS is equivalent to 147.9 BHP, typically you would see the BHP rounded up to 148 BHP in this example. 

A supercar is commonly distilled to horsepower 500 and above. The max PS being produced of hypercars are upto 2000 PS. 

In [47]:
autos['power_ps'].value_counts() # Returns the number of unique values and their count.

power_ps
0        35506
75       23188
60       15363
150      14957
140      13417
         ...  
5809         1
15020        1
5867         1
2172         1
564          1
Name: count, Length: 767, dtype: int64

In [48]:
autos['power_ps'].value_counts().sort_index() # Sorts the returned unique values' count with respect to the index.

power_ps
0        35506
1           24
2            9
3            8
4           30
         ...  
17932        1
19208        1
19211        1
19312        1
20000        1
Name: count, Length: 767, dtype: int64

In [49]:
autos['power_ps'].value_counts().sort_index().head(20) #Returns the top 20 entries in the index sorted unique values' count.

power_ps
0     35506
1        24
2         9
3         8
4        30
5        99
6        11
7        10
8         7
9         5
10       13
11       28
12       20
13        7
14       21
15       19
16       10
17        8
18       53
19       15
Name: count, dtype: int64

In [50]:
autos['power_ps'].value_counts().sort_index().tail(20) # Returns the bottom 20 entries in the index sorted unique values' count.

power_ps
13636    1
14009    1
15001    1
15017    1
15020    1
15033    1
16011    1
16051    1
16311    1
16312    1
17011    1
17019    1
17322    1
17410    1
17700    1
17932    1
19208    1
19211    1
19312    1
20000    1
Name: count, dtype: int64

We see there are min and max values in the series far extreme than nominal PS ratings for cars. 

The average min PS for most of the cars can be said to be at 150, we will thus, set our min value to be 100 to include any cars being sold 

below the average PS, and a max ceiling of 2500 PS, as an all encompassing net, even though there are no cars in the world with more than 2000 PS or above.

In [51]:
autos = autos[autos['power_ps'].between(100, 2000)]
autos['power_ps'].value_counts().sort_index()

power_ps
100      1421
101     12937
102      6378
103       950
104       134
        ...  
1993        1
1995        2
1998        3
1999        1
2000        3
Name: count, Length: 569, dtype: int64

In [52]:
autos['power_ps'].describe()

count    211325.000000
mean        156.023722
std          68.858671
min         100.000000
25%         116.000000
50%         140.000000
75%         174.000000
max        2000.000000
Name: power_ps, dtype: float64

We see the new min and max values with realistic numericals. The Standard Deviation has also decreased from 186 to 68, 
which indicates the data to be closely distributed amongst the mean, and increases certainity of the data.

In [53]:
autos['power_ps'].isnull().sum()

0

In [54]:
autos['power_ps'].kurtosis()

142.66092750189273

#### `gearbox`

In [55]:
autos['gearbox'].isnull().sum()

2718

In [56]:
autos['gearbox'].value_counts(dropna=False)

gearbox
manual       146083
automatic     62524
NaN            2718
Name: count, dtype: int64

In [57]:
autos['gearbox'].describe()

count     208607
unique         2
top       manual
freq      146083
Name: gearbox, dtype: object

In [58]:
(2645/ (143924+61097+2645)) * 100

1.2736798512996832

We see here, that 1.27% of the `gearbox` column value is NaN. And the majority is manuell, thereby we can either drop the null values or change them to the majority value. Both approach makes less difference when compared.

In [59]:
autos.isnull().sum()

date_crawled              0
name                      0
seller                    0
offer_type                0
price                     0
abtest                    0
vehicle_type          11338
registration_year         0
gearbox                2718
power_ps                  0
model                  7880
kilometer                 0
registration_month        0
fuel_type             99170
brand                     0
unrepaired_damage     26932
ad_created                0
nr_of_pictures            0
postal_code               0
last_seen                 0
dtype: int64

The above list shows the number of NaN or None values in our present dataset. With `unrepaired_damage` having the most at 26324 NaN values.

In [60]:
autos.loc[autos['gearbox'].isnull(), 'brand'].value_counts(dropna=False) 
# This code will give us the brand of the cars whose gearbox type we are missing or is set as NaN.

brand
volkswagen        548
bmw               392
opel              323
mercedes_benz     317
audi              266
ford              178
renault            73
peugeot            67
mazda              43
sonstige_autos     41
nissan             40
citroen            39
alfa_romeo         39
seat               37
fiat               36
volvo              33
porsche            28
toyota             27
honda              23
kia                20
hyundai            18
mitsubishi         18
mini               17
skoda              17
chrysler           13
chevrolet          10
rover               9
subaru              8
suzuki              7
smart               7
saab                6
jeep                4
daihatsu            3
daewoo              3
land_rover          2
lancia              2
dacia               2
trabant             1
jaguar              1
Name: count, dtype: int64

Based on internet research it seems the car brands like Volkswagen, BMW, Opel have predominatly manuel transmission or gearbox.

In [61]:
autos.loc[autos['brand'] == 'volkswagen', 'gearbox'] = 'manual'
autos.loc[autos['brand'] == 'bmw', 'gearbox'] = 'manual'
autos.loc[autos['brand'] == 'opel', 'gearbox'] = 'manual'
autos.loc[autos['brand'] == 'mercedes_benz', 'gearbox'] = 'manual'
# These lines of code will convert the NaN values for car brands with Volkswagen, BMW, Opel and mercedes_benz to manuell.

In [62]:
autos.loc[autos['gearbox'].isnull(), 'brand'].value_counts()

brand
audi              266
ford              178
renault            73
peugeot            67
mazda              43
sonstige_autos     41
nissan             40
alfa_romeo         39
citroen            39
seat               37
fiat               36
volvo              33
porsche            28
toyota             27
honda              23
kia                20
hyundai            18
mitsubishi         18
skoda              17
mini               17
chrysler           13
chevrolet          10
rover               9
subaru              8
smart               7
suzuki              7
saab                6
jeep                4
daewoo              3
daihatsu            3
land_rover          2
lancia              2
dacia               2
trabant             1
jaguar              1
Name: count, dtype: int64

In [63]:
autos['gearbox'].value_counts(dropna=False)

gearbox
manual       188144
automatic     22043
NaN            1138
Name: count, dtype: int64

In [64]:
(1098/ (143924+61097+1098)) * 100

0.5327019828351583

We now see the NaN values as dropped from 1.27% of total elements in `gearbox` column to 0.53%.

We have converted the brands with the highest NaN values in the `gearbox` column to manual.

We will keep the rest of the NaN values to be either changed or dropped at a later time, after gathering other details.

#### `vehicle_type`

In [65]:
autos['vehicle_type'].count() # returns the count of the elements in the column.

199987

In [66]:
autos['vehicle_type'].value_counts(dropna = False)

vehicle_type
limousine     69095
kombi         54350
bus           21826
cabrio        16968
coupe         15342
suv           12530
NaN           11338
kleinwagen     8538
andere         1338
Name: count, dtype: int64

In [67]:
(37869/333659) *(100)

11.349611429633248

11% NaN values in `vehicle_type`

In [68]:
# let's find out which brands have most NaN values for vehicle_type column.
autos.loc[autos['vehicle_type'].isnull(), 'brand'].value_counts(dropna = False)

brand
volkswagen        2598
bmw               1605
audi              1407
opel              1326
mercedes_benz     1077
ford               750
renault            363
seat               223
peugeot            211
mazda              188
fiat               171
alfa_romeo         133
nissan             106
mitsubishi         105
volvo              103
honda              101
skoda               96
mini                95
citroen             90
toyota              85
hyundai             85
kia                 68
sonstige_autos      65
chrysler            55
chevrolet           37
porsche             34
subaru              24
jeep                23
rover               23
suzuki              22
saab                17
daewoo              13
land_rover          10
jaguar              10
dacia                9
lancia               5
smart                4
daihatsu             1
Name: count, dtype: int64

We see there are many brands with NaN values in the `vehicle_type` column, with Volkswagen having the highest None values. It is also true that brands have multiple vehicle types, which makes replacing the NaN values with a single vehicle type inaccurate.

However, in order to have a clean dataset as much as possible and also to avoid introducing inaccurate data, we will go ahead and replace NaN values in `vehicle_type` to "other".

In [69]:
autos.loc[autos['vehicle_type'].isnull()] = 'other' # Converts all rows in vehicle_type that are true for the boolean index.
autos['vehicle_type'].value_counts(dropna = False)

vehicle_type
limousine     69095
kombi         54350
bus           21826
cabrio        16968
coupe         15342
suv           12530
other         11338
kleinwagen     8538
andere         1338
Name: count, dtype: int64

#### `model`

In [70]:
autos['model'].count()

205089

In [71]:
autos['model'].value_counts(dropna = False)

model
3er                17482
andere             14947
golf               13622
other              11338
a4                  8915
                   ...  
kalos                  1
charade                1
discovery_sport        1
sandero                1
fox                    1
Name: count, Length: 236, dtype: int64

In [72]:
autos['model'].isnull().sum()

6236

In [73]:
(autos['model'].isnull().sum())/(autos['model'].count()) * 100

3.0406311406267523

We see here, that 3.8% of the total values are NaN values in `model`.

We will go ahead and convert the NaN values of `model` to other, in the same way it was done to the `vehicle_type` column.

In [74]:
autos.loc[autos['model'].isnull()] = 'other'
autos['model'].isnull().sum()

0

#### `kilometer`

In [75]:
autos['kilometer'].count()

211325

In [76]:
autos['kilometer'].describe()

count     211325
unique        14
top       150000
freq      128239
Name: kilometer, dtype: int64

In [77]:
autos['kilometer'].value_counts(dropna = False)

kilometer
150000    128239
125000     18679
other      17574
100000      7861
90000       6461
80000       5738
70000       5279
60000       4747
50000       4172
40000       3553
30000       3324
20000       3074
5000        1615
10000       1009
Name: count, dtype: int64

It is possible that there are vehicles that have driven to the max value, shown here. 

#### `fuel_type`

In [78]:
autos['fuel_type'].count()

122301

In [79]:
autos['fuel_type'].value_counts(dropna = False)

fuel_type
benzine    104683
NaN         89024
other       17603
electro        15
Name: count, dtype: int64

The NaN values will be replaced with other, just as the previous few columns.

In [80]:
autos.loc[autos['fuel_type'].isnull()] = 'other'
autos['fuel_type'].isnull().sum()

0

#### `registration_month`

In [81]:
autos['registration_month'].count()

211325

In [82]:
autos['registration_month'].describe()

count     211325
unique        14
top        other
freq      106598
Name: registration_month, dtype: object

In [83]:
autos['registration_month'].value_counts(dropna = False)

registration_month
other    106598
3         11601
6         10146
4          9700
5          9367
7          9275
10         7903
9          7459
1          7205
11         7130
8          7008
12         7004
2          6709
0          4220
Name: count, dtype: int64

In [84]:
autos['registration_month'].info()

<class 'pandas.core.series.Series'>
Index: 211325 entries, 1 to 371527
Series name: registration_month
Non-Null Count   Dtype 
--------------   ----- 
211325 non-null  object
dtypes: object(1)
memory usage: 3.2+ MB


In [93]:
autos['registration_month'].value_counts(dropna = False)

registration_month
0.0    110818
3       11601
6       10146
4        9700
5        9367
7        9275
10       7903
9        7459
1        7205
11       7130
8        7008
12       7004
2        6709
Name: count, dtype: int64

Registration month will be left alone for the time being, we will re-visit at the end of the analysis.

### Average price of cars based on Brand/Model.

In [108]:
car_brands = autos['brand'].value_counts().head(5).index
car_brands# this will store the list of car brands.

Index(['other', 'bmw', 'volkswagen', 'mercedes_benz', 'audi'], dtype='object', name='brand')

In [112]:
avg_price_per_brand = {}

for i in car_brands:
        select_rows = autos[autos['brand'] == i]
        break
select_rows
     #   avg_price = select_rows['price'].mean()
    #    avg_price_per_brand[i] = avg_price
#avg_price_per_brand

date_crawled   name seller offer_type  price abtest vehicle_type  \
1             other  other  other      other  other  other        other   
2             other  other  other      other  other  other        other   
9             other  other  other      other  other  other        other   
11            other  other  other      other  other  other        other   
13            other  other  other      other  other  other        other   
...             ...    ...    ...        ...    ...    ...          ...   
371513        other  other  other      other  other  other        other   
371517        other  other  other      other  other  other        other   
371519        other  other  other      other  other  other        other   
371525        other  other  other      other  other  other        other   
371526        other  other  other      other  other  other        other   

       registration_year gearbox power_ps  model kilometer registration_month  \
1                  other   other    other  other     other                0.0   
2                  other   other    other  other     other                0.0   
9                  other   other    other  other     other                0.0   
11                 other   other    other  other     other                0.0   
13                 other   other    other  other     other                0.0   
...                  ...     ...      ...    ...       ...                ...   
371513             other   other    other  other     other                0.0   
371517             other   other    other  other     other                0.0   
371519             other   other    other  other     other                0.0   
371525             other   other    other  other     other                0.0   
371526             other   other    other  other     other                0.0   

       fuel_type  brand unrepaired_damage ad_created nr_of_pictures  \
1          other  other             other      other          other   
2          other  other             other      other          other   
9          other  other             other      other          other   
11         other  other             other      other          other   
13         other  other             other      other          other   
...          ...    ...               ...        ...            ...   
371513     other  other             other      other          other   
371517     other  other             other      other          other   
371519     other  other             other      other          other   
371525     other  other             other      other          other   
371526     other  other             other      other          other   

       postal_code last_seen  
1            other     other  
2            other     other  
9            other     other  
11           other     other  
13           other     other  
...            ...       ...  
371513       other     other  
371517       other     other  
371519       other     other  
371525       other     other  
371526       other     other  

[106598 rows x 20 columns]

In [113]:
autos.head()

date_crawled                                               name  \
1                other                                              other   
2                other                                              other   
5  2016-04-04 17:36:23  BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...   
6  2016-04-01 20:48:51                        Peugeot_206_CC_110_Platinum   
8  2016-04-04 23:42:13               Ford_C___Max_Titanium_1_0_L_EcoBoost   

    seller offer_type  price   abtest vehicle_type registration_year gearbox  \
1    other      other  other    other        other             other   other   
2    other      other  other    other        other             other   other   
5  private      offer    650     test    limousine              1995  manual   
6  private      offer   2200     test       cabrio              2004  manual   
8  private      offer  14500  control          bus              2014  manual   

  power_ps    model kilometer registration_month fuel_type    brand  \
1    other    other     other                0.0     other    other   
2    other    other     other                0.0     other    other   
5      102      3er    150000                 10   benzine      bmw   
6      109  2_reihe    150000                  8   benzine  peugeot   
8      125    c_max     30000                  8   benzine     ford   

  unrepaired_damage           ad_created nr_of_pictures postal_code  \
1             other                other          other       other   
2             other                other          other       other   
5               yes  2016-04-04 00:00:00              0       33775   
6                no  2016-04-01 00:00:00              0       67112   
8               NaN  2016-04-04 00:00:00              0       94505   

             last_seen  
1                other  
2                other  
5  2016-04-06 19:17:07  
6  2016-04-05 18:18:39  
8  2016-04-04 23:42:13

This notebook will be destroyed/deleted at the time.